### PR TITLE
[SPARK-22222][core] Fix the ARRAY_MAX in BufferHolder and add a test

### PR DESCRIPTION
**Pull request by liufengdb**
## What changes were proposed in this pull request?

We should not break the assumption that the length of the allocated byte array is word rounded:  
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java#L170
So we want to use `Integer.MAX_VALUE - 15` instead of `Integer.MAX_VALUE - 8` as the upper bound of an allocated byte array. 

cc: @srowen @gatorsmile 
## How was this patch tested?

Since the Spark unit test JVM has less than 1GB heap, here we run the test code as a submit job, so it can run on a JVM has 4GB memory. 

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

File: common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java [hunk 1]
```diff
@@ -40,6 +40,15 @@ public static int roundNumberOfBytesToNearestWord(int numBytes) {
     }
   }
 
+  // Some JVMs can't allocate arrays of length Integer.MAX_VALUE; actual max is somewhat smaller.
+  // Be conservative and lower the cap a little.
+  public static int MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
```
Comment (gatorsmile) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java]: `- 15`

File: common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java [hunk 2]
```diff
@@ -40,6 +40,15 @@ public static int roundNumberOfBytesToNearestWord(int numBytes) {
...
+
+  // Use this value if the allocated byte arrays are used to store other types rather than bytes.
+  public static int maxWordRoundedArrayLength() {
+    return roundNumberOfBytesToNearestWord(MAX_ARRAY_LENGTH - 8);
```
Comment (srowen) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java]: Rounding won't do; it has to round _down_. I see that's why you wrote -8 here but it's a little confusing as -7 works more directly. And then, why bother? just write -15 in the constant above and explain.

File: common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java (hunk 1 above)
Comment (srowen) [inline common/unsafe/src/main/java/org/apache/spark/unsafe/array/ByteArrayMethods.java]: I get declaring a constant though it doesn't just pertain to byte arrays. I couldn't find a good place for it. I don't know if its reused so much that it needs this

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolderSuite.scala [hunk 3]
```diff
@@ -37,3 +43,54 @@ class BufferHolderSuite extends SparkFunSuite {
     assert(e.getMessage.contains("exceeds size limitation"))
   }
 }
+
+// A test for growing the buffer holder to nearly 2GB. Due to the heap size limitation of the Spark
+// unit tests JVM, the actually test code is running as a submit job.
+class BufferHolderSparkSubmitSuite
```
Comment (srowen) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolderSuite.scala]: Separate file?

**Comment by gatorsmile** [comment]
LGTM

**Comment by kiszk** [comment]
LGTM

**Comment by gatorsmile** [comment]
Thanks! Merged to master.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java [hunk 4]
```diff
@@ -36,9 +37,7 @@
  */
 public class BufferHolder {
 
-  // Some JVMs can't allocate arrays of length Integer.MAX_VALUE; actual max is somewhat
-  // smaller. Be conservative and lower the cap a little.
-  private static final int ARRAY_MAX = Integer.MAX_VALUE - 8;
+  private static final int ARRAY_MAX = ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH;
```
Comment (srowen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/BufferHolder.java]: Ok, these constants became redundant. No big deal. We don't need to rush this though

**Pull request closed without merging**